<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [55]:
import pandas as pd
import psycopg2

In [57]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [58]:
# текст запроса
query_3_1 = f''' SELECT COUNT(id) "Количество вакансий"
                 FROM public.VACANCIES
'''


In [59]:
# результат запроса
df_3_1 = pd.read_sql_query(query_3_1, connection)
df_3_1

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\4045908245.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество вакансий
0                49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [60]:
# текст запроса
query_3_2 = f''' SELECT COUNT(id) "Количество работодателей"
                 FROM public.EMPLOYERS
'''

In [61]:
# результат запроса
df_3_2 = pd.read_sql_query(query_3_2, connection)
df_3_2

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\533540996.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество работодателей
0                     23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [62]:
# текст запроса
query_3_3 = f''' SELECT COUNT(id) "Количество регионов"
                 FROM public.AREAS
'''

In [63]:
# результат запроса
df_3_3 = pd.read_sql_query(query_3_3, connection)
df_3_3

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\1395976625.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество регионов
0                 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [64]:
# текст запроса
query_3_4 = f''' SELECT COUNT(id) "Количество сфер деятельности"
                 FROM public.INDUSTRIES
'''

In [65]:
# результат запроса
df_3_4 = pd.read_sql_query(query_3_4, connection)
df_3_4

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\854197215.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество сфер деятельности
0                           294

***

In [66]:
# выводы по предварительному анализу данных

##### ПРЕДВАРИТЕЛЬНЫЙ АНАЛИЗ #####
Предварительный анализ данных показал, что *вакансий*  в базе данных HeadHunter достаточно много (**49 197**). Также есть возможность выбора из достаточно большого количства *работодателей* (**23 501**) и *сфер деятельности* (**294**). Кроме того, данные из таблицы имеют широкий географический охват, **1 362** *региона*. 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [67]:
# текст запроса
query_4_1 = f''' SELECT 
                    COUNT(vac.id) cnt,
                    a.name area
                 FROM public.VACANCIES vac
                 JOIN public.AREAS a ON vac.area_id = a.id
                 GROUP BY a.id
                 ORDER BY 1 DESC
                 LIMIT 15
'''

In [68]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\620650199.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt             area
0   5333           Москва
1   2851  Санкт-Петербург
2   2112            Минск
3   2006      Новосибирск
4   1892           Алматы
5   1698     Екатеринбург
6   1670  Нижний Новгород
7   1415           Казань
8   1301        Краснодар
9   1144           Самара
10  1131   Ростов-на-Дону
11  1063          Воронеж
12   851          Ташкент
13   847       Красноярск
14   840       Нур-Султан

In [69]:
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

fig = px.bar(
    data_frame=df_4_1,
    x='area',
    y='cnt',
    title='Зависимость количества вакансий от региона'
)
fig.show()
fig.write_html('1.html');

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [70]:
# текст запроса
query_4_2 = f''' SELECT 
                    COUNT(vac.id) cnt
                 FROM public.VACANCIES vac
                 WHERE (vac.salary_from IS NOT NULL) OR (vac.salary_to IS NOT NULL)
'''

In [71]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\3746829199.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [72]:
# текст запроса
query_4_3 = f''' SELECT 
                    AVG(vac.salary_from) avg_min,
                    AVG(vac.salary_to) avg_max
                FROM public.VACANCIES vac
'''

In [73]:
# результат запроса
df_4_3 = round(pd.read_sql_query(query_4_3, connection))
df_4_3

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\569676163.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



avg_min   avg_max
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [74]:
# текст запроса
query_4_4 = f''' SELECT 
                    COUNT(vac.id),
                    vac.schedule,
                    vac.employment
                FROM public.VACANCIES vac
                GROUP BY vac.schedule, vac.employment
                ORDER BY 1 DESC
'''

In [75]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\2028972612.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

In [76]:
fig = px.bar(
    data_frame=df_4_4,
    y='employment',
    x='count',
    color='schedule', 
    barmode="group",
    title='Зависимость количества вакансий от рабочего графика и типа занятости'
)
fig.update_traces(dict(marker_line_width=0))
fig.update_layout(
    xaxis_title="Количество вакансий"
)
fig.show()
fig.write_html('2.html');

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [77]:
# текст запроса
query_4_5 = f''' SELECT 
                    vac.experience,
                    COUNT(vac.id)
                FROM public.VACANCIES vac
                GROUP BY vac.experience
                ORDER BY 2
'''

In [78]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\210463167.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

In [79]:
fig = px.bar(
    data_frame=df_4_5,
    x='experience',
    y='count',
    title='Зависимость количества вакансий от опыта работы'
)
fig.show()
fig.write_html('3.html');

In [80]:
# выводы по детальному анализу вакансий

##### ДЕТАЛЬНЫЙ АНАЛИЗ #####
Детальный анализ данных показал, что:
- Больше всего вакансий представлено в **Москве**. В два раза меньше вакансий представлено в других крупных городах. Можно в дальнейшем исследовать взаимосвязь между количеством вакансий и количеством трудоспособного населения в регионе. 
- Примерно половина вакансий имеет заполненые поля с зарплатой "от" и/или "до". Среднее значение "от" - **71 065**, "до" - **110 537**.
- Что касается работы и типа занятости, то подавляющее большинство вакансий предполагает **полную занятость** и **полный день**. Примерно в четыре раза меньше вакансий с полной занятостью и удаленной работой. Меньше всего вакансий предполагает проектную работу.
- Больше всего работадатели заинтересованы в работниках с оптом работы **от 1 года до 3 лет**. Потребность в опытных сотрудниках (стаж больше 6 лет) наименьшая.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [81]:
# текст запроса
query_5_1 = f''' SELECT 
                    emp.name,
                    COUNT(vac.id)
                FROM public.EMPLOYERS emp
                JOIN public.VACANCIES vac ON vac.employer_id = emp.id
                GROUP BY emp.name
                ORDER BY 2 DESC
                LIMIT 5
'''

In [82]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\1147638897.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

In [83]:
fig = px.bar(
    data_frame=df_5_1,
    x='name',
    y='count',
    title='Топ-5 работодателей'
)
fig.show()
fig.write_html('4.html');

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [84]:
# текст запроса
query_5_2 = f''' SELECT 
                    ar.name,
                    COUNT(distinct emp.id) as COUNT_employers,
                    COUNT(distinct vac.id) as COUNT_vacancies
                FROM public.AREAS ar
                LEFT JOIN public.EMPLOYERS emp ON ar.id = emp.area 
                LEFT JOIN public.VACANCIES vac ON vac.area_id = ar.id
                WHERE vac.id IS NULL
                GROUP BY 1
                ORDER BY 2 DESC
                LIMIT 10
'''

In [85]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\2306256317.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count_employers  count_vacancies
0                 Россия              410                0
1              Казахстан              207                0
2     Московская область               75                0
3     Краснодарский край               19                0
4               Беларусь               18                0
5     Ростовская область               18                0
6            Азербайджан               17                0
7  Нижегородская область               16                0
8   Республика Татарстан               16                0
9             Узбекистан               15                0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [86]:
# текст запроса
query_5_3 = f''' SELECT 
                    emp.name,
                    COUNT(distinct ar.id) as COUNT_areas
                FROM public.EMPLOYERS emp     
                LEFT JOIN public.VACANCIES vac ON vac.employer_id = emp.id
                LEFT JOIN public.AREAS ar ON ar.id = vac.area_id 
                GROUP BY 1
                ORDER BY 2 DESC
'''

In [87]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\133400711.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count_areas
0                      Яндекс          181
1                  Ростелеком          152
2                  Спецремонт          116
3      Поляков Денис Иванович           88
4                   ООО ЕФИН            71
...                       ...          ...
23170                  СДЕЛКА            0
23171            Alandr Group            0
23172                     СДК            0
23173      Lemon Land Lombard            0
23174                  Махина            0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [88]:
# текст запроса
query_5_4 = f''' SELECT 
                    COUNT(emp.name)
                FROM public.EMPLOYERS emp     
                LEFT JOIN public.EMPLOYERS_INDUSTRIES ei ON ei.employer_id = emp.id
                WHERE ei.industry_id IS NULL
'''

In [89]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\1045119256.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [90]:
# текст запроса
query_5_5 = f''' SELECT 
                    emp.name,
                    COUNT(ei.industry_id)
                FROM public.EMPLOYERS emp     
                LEFT JOIN public.EMPLOYERS_INDUSTRIES ei ON ei.employer_id = emp.id
                GROUP BY emp.name
                HAVING COUNT(ei.industry_id) = 4
                ORDER BY 1
                OFFSET 2
                LIMIT 1
'''

In [91]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\4025176131.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [92]:
# текст запроса
query_5_6 = f''' SELECT 
                    COUNT(emp.id)
                FROM public.EMPLOYERS emp     
                LEFT JOIN public.EMPLOYERS_INDUSTRIES ei ON ei.employer_id = emp.id
                LEFT JOIN public.INDUSTRIES ind ON ei.industry_id = ind.id
                WHERE ind.name = 'Разработка программного обеспечения'
'''                

In [93]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\2713946032.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [94]:
# код для получения списка городов-милионников
import urllib.parse as req

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
cities = tuple(pd.read_html(url)[0]['Город'])
display(cities)

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Красноярск',
 'Нижний Новгород',
 'Челябинск',
 'Уфа',
 'Самара',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [95]:
# текст запроса
query_5_7 = f''' ((SELECT 
                    ar.name,
                    COUNT(vac.id) cnt
                FROM public.VACANCIES vac     
                JOIN public.AREAS ar ON vac.area_id = ar.id
                JOIN public.EMPLOYERS emp ON emp.id = vac.employer_id
                WHERE emp.name = 'Яндекс'
                GROUP BY ar.name
                HAVING ar.name IN {cities})
                UNION ALL
                (SELECT 
                    'total',
                    COUNT(vac.id) cnt
                FROM public.VACANCIES vac     
                JOIN public.AREAS ar ON vac.area_id = ar.id
                JOIN public.EMPLOYERS emp ON emp.id = vac.employer_id
                WHERE emp.name = 'Яндекс' AND ar.name IN {cities}))
                ORDER BY 2
                
'''

In [96]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\2595257991.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            total  485

***

In [97]:
fig = px.bar(
    data_frame=df_5_7,
    x='name',
    y='cnt',
    title='Количество вакансий Яндекса в городах-миллионниках'
)
fig.show()
fig.write_html('5.html');

In [98]:
# выводы по анализу работодателей

##### АНАЛИЗ РАБОТОДАТЕЛЕЙ #####
Анализ работадателей показал, что:
- Больше всего вакансий предоставлены крупными российскими компаниями, **"Яндекс"** (почти в пять раз больше, чем у следующей компании), **"Ростелеком"**, **"Тинькофф"**, **"СБЕР"**, **"Газпром нефть"**. 
- Есть регионы в которых зарегистрированы работадатели, но нет вакансий. Например, **Россия, Казахстан** и т.п. Скорее всего, это связано с тем, что регион указан слишком широко. Есть компании, которые публикуют свои вакансии в большом количестве регионов. **Яндекс** представлен в 181 регионах, **Ростелеком** - в 152 регионах.
- 8419 компаний не указали в своих вакансиях сферу деятельности. Есть компании, у которых указано несколько сфер деятельности. У 3553 работадателей **"Разработка программного обеспечения"** указана в качестве сферы деятельности.
- Во всех **городах-миллионниках** вакансии Яндекса представлены. Ожидаемо, Москва оказалась лидером по количеству вакансий, но разрыв между столицеий и другими городами-миллионниками не слишком большой.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [99]:
# текст запроса
query_6_1 = f''' SELECT 
                    COUNT(vac.id)
                FROM public.VACANCIES vac    
                WHERE (LOWER(vac.name) LIKE '%data%') OR (LOWER(vac.name) LIKE '%данн%')
''' 

In [100]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\1281472032.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [101]:
# текст запроса
query_6_2 = f''' SELECT 
                    COUNT(vac.id)
                FROM public.VACANCIES vac    
                WHERE ((LOWER(vac.name) LIKE '%data scientist%') OR (LOWER(vac.name) LIKE '%data science%') 
                OR (LOWER(vac.name) LIKE '%исследователь данных%') 
                OR ((LOWER(vac.name) LIKE '%ml%') AND NOT (LOWER(vac.name) LIKE '%html%'))
                OR (LOWER(vac.name) LIKE '%machine learning%') OR (LOWER(vac.name) LIKE '%машинн%обучен%'))
                AND ((LOWER(vac.name) LIKE '%junior%')
                OR (LOWER(vac.experience) LIKE '%нет опыта%') OR (LOWER(vac.employment) LIKE '%стажировка%'))
'''

In [102]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\118162426.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [103]:
# текст запроса
query_6_3 = f''' WITH t AS
                (SELECT 
                    *
                FROM public.VACANCIES vac    
                WHERE ((LOWER(vac.name) LIKE '%data scientist%') OR (LOWER(vac.name) LIKE '%data science%') 
                OR (LOWER(vac.name) LIKE '%исследователь данных%') 
                OR ((vac.name ILIKE '%ML%') AND NOT (vac.name ILIKE '%HTML%') )
                OR (LOWER(vac.name) LIKE '%machine learning%') OR (LOWER(vac.name) LIKE '%машинн%обучен%')))
                SELECT 
                    COUNT(id)
                FROM t
                WHERE (key_skills LIKE '%SQL%') OR (key_skills LIKE '%postgres%')
''' 

In [104]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\3310962512.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [105]:
# текст запроса
query_6_4 = f''' WITH t AS
                (SELECT 
                    *
                FROM public.VACANCIES vac    
                WHERE ((LOWER(vac.name) LIKE '%data scientist%') OR (LOWER(vac.name) LIKE '%data science%') 
                OR (LOWER(vac.name) LIKE '%исследователь данных%') 
                OR ((vac.name ILIKE '%ML%') AND NOT (vac.name ILIKE '%HTML%') )
                OR (LOWER(vac.name) LIKE '%machine learning%') OR (LOWER(vac.name) LIKE '%машинн%обучен%')))
                SELECT 
                    COUNT(id)
                FROM t
                WHERE (key_skills LIKE '%Python%')
'''

In [106]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\614304167.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [107]:
# текст запроса
query_6_5 = f''' WITH t AS
                (SELECT 
                    *
                FROM public.VACANCIES vac    
                WHERE ((LOWER(vac.name) LIKE '%data scientist%') OR (LOWER(vac.name) LIKE '%data science%') 
                OR (LOWER(vac.name) LIKE '%исследователь данных%') 
                OR ((vac.name LIKE '%ML%') AND NOT (vac.name LIKE '%HTML%') )
                OR (LOWER(vac.name) LIKE '%machine learning%') OR (LOWER(vac.name) LIKE '%машинн%обучен%')))
                SELECT 
                    ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), ''))+1), 2)
                FROM t
'''

In [108]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\4089566367.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [109]:
# текст запроса
query_6_6 = f''' WITH t AS
                (SELECT 
                    *
                FROM public.VACANCIES vac    
                WHERE ((LOWER(vac.name) LIKE '%data scientist%') OR (LOWER(vac.name) LIKE '%data science%') 
                OR (LOWER(vac.name) LIKE '%исследователь данных%') 
                OR ((vac.name LIKE '%ML%') AND NOT (vac.name LIKE '%HTML%') )
                OR (LOWER(vac.name) LIKE '%machine learning%') OR (LOWER(vac.name) LIKE '%машинн%обучен%')))
                SELECT 
                    ROUND(AVG((coalesce(salary_to,salary_from)+coalesce(salary_from,salary_to))/2))
                FROM t
                WHERE ((salary_to IS NOT NULL) OR (salary_from IS NOT NULL)) AND (experience = 'От 3 до 6 лет')
'''

In [110]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

C:\Users\Даша\AppData\Local\Temp\ipykernel_16364\2297777544.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



round
0  243115.0

***

In [111]:
connection.close()

In [112]:
# выводы по предметному анализу

##### ПРЕДМЕТНЫЙ АНАЛИЗ #####
Предметный анализ показал, что:
- **1771** вакансия из представленных имеют отношение к данным. **536** вакансий соответствуют специалистам в *Data Science*, **51** вакансия из них подходит начинающим. Далее анализ будет проводится только для вакансий Data Science.
- **229** вакансий имеют в качестве ключевого навыка *SQL* или *postgres*. Значит, данные навыки играют важную роль для соискателя в данной области. **357** вакансий имеют в качестве ключевого навыка *Python*. Иными словами, больше половины вакансий в Data Science предполагают знания этого языка программирования. В среднем количество ключевых навыков равно **6.41**, что говорит нам о достаточно высоких требованиях к специалистам в Data Science
- Средняя зарплата специалиста в Data Science с опытом работы от 3 до 6 лет равна **243 115**.

# Общий вывод по проекту

In [113]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

#### **ОБЩИЙ ВЫВОД ПО ПРОЕКТУ** ####
Исследования показали, что в базе данных **Head Hunter** есть вакансии в разных регионах, в разных сферах. Специалисты требуются с разными навыками и с разным опытом. В Москве представлено много вакансий, больший выбор, чем в других городах. По-прежнему наиболее распространенными являются вакансии с полным рабочим днем и полной занятостью. Но есть и часть вакансий, предполагающих удаленную работу, что подходит жителям перферии.
Специалисты в **Data Science** являются востребованными и хорошо оплачиваемыми. Но для конкурентной борьбы в этом сегменте рынка труда необходимо обладать несколькими ключевыми навыками (Python, SQL и т.д.)
В зависимости от поставленной задачи можно продолжить исследования. Интересно найти сферу, регион, компанию, где зарплата наибольшая, наименьшая. В каких специалистах нуждается рынок труда? Какие ключевые навыки являются самыми востребованными или высокооплачиваемыми? Есть ли взаимосвязь между сферами деятельности и графиком работы, опытом, типом трудоустройства? 